Migration to iced's "pure" widget design

fengalin opened the pull request:
This is a proposal for a migration to iced's "pure" widget design. See https://docs.rs/iced/latest/iced/pure/index.html.
It targets iced current main branch.

The second commit suggests slight improvements, which are independent of the migration.

Fixes https://github.com/iced-rs/iced_audio/issues/23

fengalin commented:
Updated to latest `iced` release.

greatest-ape commented:
Nice! I'll try to have a look when I have time.

fengalin commented:
I thought current state was good enough for supporting `Style`s but it seems it's not! I'm working on an application for which I want the widgets in a `Component` and the old `Style` handling gets in the way. I'm looking into this issue ATM, so I'll probably push some fixes in the coming days.

greatest-ape commented:
Alright, great!

fengalin commented:
Style handling migrated!

FWIW, my issue with the integration of `iced_audio` widgets in a `Component` didn't have to do with theme / style handling. I needed to restrict the `Component`'s `Renderer` implementation to `iced::Renderer` since `iced_audio` only supports this `Renderer` currently. Theme & Style handling needed a migration anyway.

IMO this repository would require a clippy cleanup :) I can do this in another PR if this one gets merged.

greatest-ape commented:
Nice!

on_drag_start is necessary to be able to call begin_automate for vst plugins. It could be renamed, if course.

fengalin commented:
> on_drag_start is necessary to be able to call begin_edit for vst plugins. It could be renamed, if course.

All right! I'll add it back in and see if I can find another name.

greatest-ape commented on Cargo.toml:
```diff
@@ -25,9 +25,7 @@ members = [
 [dependencies]
-iced_native = "0.5"
-iced_graphics = { version = "0.3", features = ["canvas"] }
+iced = { version = "0.5.2", features = ["canvas"] }
```
I can't get this to build with OctaSine unless `default-features = false` is set for iced and iced_audio features for wgpu and glow are added, like:

```
[features]
default = ["wgpu"]
glow = ["iced/glow"]
wgpu = ["iced/wgpu"]
```

greatest-ape replied:
Oh, the canvas feature can definitely be kept! It is just that iced defaults to enabling wgpu, which makes widgets non-compatible with the glow backend. I don’t have the exact error message available now.

fengalin replied:
Ok, I see. I'll update accordingly.

fengalin commented:
> on_drag_start could be called on_press?

I wanted to propose `on_grab` as I thought it would be symmetrical with `on_release`. What do you think? Otherwise, `on_press` is fine! :)

> Ideally, it should be implemented for all widgets. I just didn't have the need for it for any other widgets than knobs when I implemented it.

Yes, that's what I thought too. I'll had it to all widgets.

> Another thing is that it would be good if it remained possible to check if the knob is currently in the dragging state.

Yes, it's still possible. Even though the `State` isn't fully handled by the user, pure widgets can still use an internal `State`. In last commit, I renamed `dragging_status` as `is_dragging` because it is now more than a boolean, but it still tracks the dragging status.

Considering the plugins requirements you mentioned, my plan is:

- If user defines `on_release` only, call it only if a significant movement happened since the widget was "grabbed".
- If user defines both `on_grab` and `on_release`, trigger both regardless of the actual movements.
- Of course, user can define none of the optional callbacks, in which case only `on_change` is triggered when a significant movement happens.

Does this seem appropriate to you?

greatest-ape commented:
> > on_drag_start could be called on_press?
> 
> I wanted to propose `on_grab` as I thought it would be symmetrical with `on_release`. What do you think? Otherwise, `on_press` is fine! :)

on_grab sounds good to me.

> Yes, it's still possible. Even though the `State` isn't fully handled by the user, pure widgets can still use an internal `State`. In last commit, I renamed `dragging_status` as `is_dragging` because it is now more than a boolean, but it still tracks the dragging status.

I don’t think I follow you here, is it still possible for the user to decide whether to do data updates depending on if the knob is dragging?

> 
> Considering the plugins requirements you mentioned, my plan is:
> 
> * If user defines `on_release` only, call it only if a significant movement happened since the widget was "grabbed".
> * If user defines both `on_grab` and `on_release`, trigger both regardless of the actual movements.
> * Of course, user can define none of the optional callbacks, in which case only `on_change` is triggered when a significant movement happens.
> 
> Does this seem appropriate to you?

Sure, I think it would work!

fengalin commented:
> is it still possible for the user to decide whether to do data updates depending on if the knob is dragging?

The user doesn't get access to `is_dragging` in current implementation. However the `on_change` callback, in which the user is supposed to update the `NormalParam` which is fed back to the widget, is only fired when the widget is being dragged or the mouse wheel is scrolled. An external update of the `NormalParam`, such as resulting from a device event or automation, won't fire `on_change`. Does this address your concern?

greatest-ape commented:
Not 100%, but it’s ok :-)

greatest-ape commented:
Could on_grab and on_release take closures producing messages instead of just messages?

greatest-ape commented:
I'm thinking of having the option of generating messages lazily. But it's not important, it can be added if needed.

fengalin commented:
The last two commits should address all the topics we discussed.

greatest-ape commented:
Excellent! The iced dependency still needs default-features = false, though.

fengalin commented:
> The iced dependency still needs default-features = false, though.

Sorry about that. That should be fixed now.

greatest-ape commented:
Thanks!

greatest-ape commented:
> IMO this repository would require a clippy cleanup :) I can do this in another PR if this one gets merged.

Please do :-)
